lib/src/geometry/polygon.dart

```diff
@@ -0,0 +1,156 @@
+
+  /// Gives back the shape vectors multiplied by the size
+  Iterable<Vector2> get scaled {
+    if (!_cachedScaledShape.isCacheValid(<dynamic>[size])) {
+      _cachedScaledShape.updateCache(
+        () => definition?.map((p) => p.clone()..multiply(size / 2)),
+        <dynamic>[size.clone()],
+      );
+    }
+    return _cachedScaledShape.value;
+  }
+
+  final _cachedRenderPath = ShapeCache<Path>();
+
+  @override
+  void render(Canvas canvas, Paint paint) {
+    if (!_cachedRenderPath.isCacheValid(<dynamic>[position, size])) {
+      _cachedRenderPath.updateCache(
+        () => Path()
+          ..addPolygon(
+            scaled
+                .map((point) => (point +
+                        (position + size / 2) +
+                        ((size / 2)..multiply(relativePosition)))
+                    .toOffset())
+                .toList(),
+            true,
+          ),
+        <dynamic>[
+          position.clone(),
+          size.clone(),
+        ],
+      );
+    }
+    canvas.drawPath(_cachedRenderPath.value, paint);
+  }
+
+  final _cachedHitbox = ShapeCache<List<Vector2>>();
+
+  /// Gives back the vertices represented as a list of points which
+  /// are the "corners" of the hitbox rotated with [angle].
+  List<Vector2> get hitbox {
+    // Use cached bounding vertices if state of the component hasn't changed
+    if (!_cachedHitbox.isCacheValid(<dynamic>[position, size, angle])) {
+      _cachedHitbox.updateCache(
+        () {
+          return scaled
+                  .map((point) => (point + shapeCenter)
+                    ..rotate(angle, center: anchorPosition))
+                  .toList(growable: false) ??
+              [];
+        },
+        <dynamic>[shapeCenter, size.clone(), angle],
+      );
+    }
+    return _cachedHitbox.value;
+  }
+
+  /// Checks whether the polygon represented by the list of [Vector2] contains
+  /// the [point].
+  @override
+  bool containsPoint(Vector2 point) {
+    // If the size is 0 then it can't contain any points
+    if (size.x == 0 || size.y == 0) {
+      return false;
+    }
+
+    for (int i = 0; i < hitbox.length; i++) {
+      final previousNode = hitbox[i];
+      final node = hitbox[(i + 1) % hitbox.length];
+      final isOutside = (node.x - previousNode.x) * (point.y - previousNode.y) -
+              (point.x - previousNode.x) * (node.y - previousNode.y) >
+          0;
+      if (isOutside) {
+        // Point is outside of convex polygon
+        return false;
+      }
+    }
+    return true;
+  }
+
+  /// Return all [vertices] as [LineSegment]s that intersect [rect], if [rect]
+  /// is null return all [vertices] as [LineSegment]s.
+  List<LineSegment> possibleIntersectionVertices(Rect rect) {
+    final List<LineSegment> rectIntersections = [];
+    final vertices = hitbox;
+    for (int i = 0; i < vertices.length; i++) {
+      final from = vertices[i];
+      final to = vertices[(i + 1) % vertices.length];
```

Comment:
this could also use a getEdges method?

Reply:
I tried to do this:
```dart
  LineSegment getEdge(int x) {
    return LineSegment(hitbox[x], hitbox[(x + 1) % hitbox.length]);
  }
```
but it turned out to be a real performance hit, it is crazy how expensive it seems to be to create objects in dart.
So I just created this instead, it doesn't make the rest of the code as pretty, but prettier than it is currently at least.
```dart
  Vector2 getVertex(int x) {
    return hitbox[x % hitbox.length];
  }
```

Reply:
Ah, it was the hitbox getter of course that pulled it down, fixed it now.